privacy/enforcer.go

```diff
@@ -1,43 +1,241 @@
+	if err != nil {
+		return ef, err
+	}
+	ef.defaultResult = activityDefaultToDefaultResult(activity.Default)
+	ef.rules = rules
+	return ef, nil
+}
+
+func activityRulesToEnforcementRules(rules []config.ActivityRule) ([]ActivityRule, error) {
+	enfRules := make([]ActivityRule, 0)
+	for _, r := range rules {
+		cmpName, err := conditionToRuleComponentName(r.Condition.ComponentName)
+		if err != nil {
+			return nil, err
+		}
+		er := ComponentEnforcementRule{
+			allowed:       r.Allow,
+			componentName: cmpName,
+			componentType: r.Condition.ComponentType,
+		}
+		enfRules = append(enfRules, er)
+	}
+	return enfRules, nil
+}
+
+func conditionToRuleComponentName(conditions []string) ([]ScopedName, error) {
+	sn := make([]ScopedName, 0)
+	for _, condition := range conditions {
+		scope, err := NewScopedName(condition)
+		if err != nil {
+			return sn, err
+		}
+		sn = append(sn, scope)
+	}
+	return sn, nil
+}
+
+func activityDefaultToDefaultResult(activityDefault *bool) ActivityResult {
+	if activityDefault == nil {
+		// if default is unspecified, the hardcoded default-default is true.
+		return ActivityAllow
+	} else if *activityDefault {
+		return ActivityAllow
+	}
+	return ActivityDeny
 }
 
-// NilPolicyEnforcer implements the PolicyEnforcer interface but will always return false.
-type NilPolicyEnforcer struct{}
+func (e ActivityControl) Allow(activity Activity, target ScopedName) ActivityResult {
+	plan, planDefined := e.plans[activity]
+
+	if !planDefined {
+		return ActivityAbstain
+	}
+
+	return plan.Allow(target)
+}
+
+type ActivityPlan struct {
+	defaultResult ActivityResult
+	rules         []ActivityRule
+}
+
+func (p ActivityPlan) Allow(target ScopedName) ActivityResult {
+	for _, rule := range p.rules {
+		result := rule.Allow(target)
+		if result == ActivityDeny || result == ActivityAllow {
+			return result
+		}
+	}
+	return p.defaultResult
+}
 
-// CanEnforce is hardcoded to always return false.
-func (NilPolicyEnforcer) CanEnforce() bool {
-	return false
+type ActivityRule interface {
+	Allow(target ScopedName) ActivityResult
 }
 
-// ShouldEnforce is hardcoded to always return false.
-func (NilPolicyEnforcer) ShouldEnforce(bidder string) bool {
-	return false
+type ComponentEnforcementRule struct {
+	componentName []ScopedName
+	componentType []string
+	// include gppSectionId from 3.5
+	// include geo from 3.5
+	allowed bool
 }
 
-// EnabledPolicyEnforcer decorates a PolicyEnforcer with an enabled flag.
-type EnabledPolicyEnforcer struct {
-	Enabled        bool
-	PolicyEnforcer PolicyEnforcer
+func (r ComponentEnforcementRule) Allow(target ScopedName) ActivityResult {
+	if len(r.componentName) == 0 && len(r.componentType) == 0 {
+		return ActivityAbstain
+	}
+
+	componentNameFound := false
+	if len(r.componentName) == 0 {
+		componentNameFound = true
+	}
+	for _, scope := range r.componentName {
+		if strings.EqualFold(scope.Scope, target.Scope) &&
+			(strings.EqualFold(scope.Name, target.Name) || scope.Name == "*") {
```

Comment:
The compiler would hopefully optimize it anyway, but you could put the `scope.Name == "*"` first in the or since it is cheaper to compute than a longer string compare.